bibtex-actions.el

```diff
@@ -72,26 +133,56 @@
                "BibTeX entries: "
                (lambda (string predicate action)
                  (if (eq action 'metadata)
-                     '(metadata
-                       ;; TODO (affixation-function . bibtex-actions--affixation)
+                     `(metadata
+                       ,(when bibtex-actions-rich-ui
+                          '(affixation-function . bibtex-actions--affixation))
                        (category . bibtex))
                    (complete-with-action action candidates string predicate))))))
     (cl-loop for choice in chosen
-             collect (cdr (assoc choice candidates)))))
+             ;; collect citation keys of selected candidate(s)
+             collect (cdr (assoc "=key=" (cdr (assoc choice candidates)))))))
 
 (defun bibtex-actions--get-candidates ()
-  "Return all keys from 'bibtex-completion-candidates'."
+  "Propertize the candidates from 'bibtex-completion-candidates'."
   (cl-loop
    for candidate in (bibtex-completion-candidates)
    collect
+   (let* ((pdf (if (assoc "=has-pdf=" (cdr candidate)) " has:pdf"))
+          (note (if (assoc "=has-note=" (cdr candidate)) "has:note"))
+          (link (if (assoc "doi" (cdr candidate)) "has:link"))
+          (add (s-join " " (list pdf note link))))
    (cons
     ;; Here use one string for display, and the other for search.
     ;; The candidate string we use is very long, which is a bit awkward
     ;; when using TAB-completion style multi selection interfaces.
     (propertize
-     (car candidate) 'display (bibtex-completion-format-entry
-     candidate (1- (frame-width)))) ; allow this to be configurable?
-    (cdr (assoc "=key=" candidate)))))
+     (s-append add (car candidate)) 'display (bibtex-completion-format-entry
+     candidate (1- (frame-width))))
+    (cdr candidate)))))
+
+(defun bibtex-actions--affixation (cands)
+  "Add affixes to CANDS."
+  (cl-loop
+   for candidate in cands
+   collect
+   (let ((pdf (if (string-match "has:pdf" candidate)
+                  (car (cdr (assoc 'pdf bibtex-actions-icon)))
+                (cdr (cdr (assoc 'pdf bibtex-actions-icon)))))
+         (link (if (string-match "has:link" candidate)
+                  (car (cdr (assoc 'link bibtex-actions-icon)))
+                (cdr (cdr (assoc 'link bibtex-actions-icon)))))
+         (note
+          (if (string-match "has:note" candidate)
+                  (car (cdr (assoc 'note bibtex-actions-icon)))
+                (cdr (cdr (assoc 'note bibtex-actions-icon))))))
+   (list candidate (concat
+                    (s-join bibtex-actions-icon-separator
+                            (list pdf note link))"	")"	XYZ"))))
+
+(defun bibtex-actions--make-suffix (entry)
+  "Create the formatted ENTRY suffix string for the 'rich-ui'."
+    ;;TODO make use of 'bibtex-completion' so as simple and flexible as possible.
```

Comment:
Is there a way to advise, or otherwise hack, rather than recreate that code, [bibtex-completion-format-entry](https://github.com/tmalsburg/helm-bibtex/blob/ca09076c3d6e51cc4ffe208c8176fccf3710fcc6/bibtex-completion.el#L851) to instead generate the suffix?

Reply:
What're you trying to make `bibtex-actions--make-suffix` do? Not include a pdf, note and link symbol? If so I think the easiest approach would be to modify `bibtex-completion-display-formats` so that it doesn't contain those formats.

Reply:
> What're you trying to make `bibtex-actions--make-suffix` do? Not include a pdf, note and link symbol? If so I think the easiest approach would be to modify `bibtex-completion-display-formats` so that it doesn't contain those formats.

No; I'm not worried about that.

The idea I had for the suffix was it would be for secondary metadata related to the candidate.

So main display is primary (author, date, title, for example), and suffix is things perhaps like tags, type, etc.

I was assuming people would want to be able to configure it, just like they can now for the main display using `bibtex-completion-display-formats`.

What you ask seems to suggest using that variable (and function) to format the suffix, but formatting authors, titles, dates can be messy. I'd really rather leave that to `bibtex-completion` if possible.

Reply:
Ah, I see. I don't think advice would work very well here. Honestly I don't really see the need for a suffix, you can probably push all of that into main. I'd say the best approach would be to explain your intentions to the bibtex maintainer and ask if he'd be willing to change to get the behaviour you need. Otherwise, yeah, re-implementation is the only approach I can think of.

Reply:
Yeah, what I was thinking.

I'll hold off awhile on asking, as I know he's really busy, and as you note, it's not critical.

Can you confirm this works as expected on your end when you get a chance, including the actual commands (open and such)?

I'm seeing some inconsistent behavior on my end, but I suspect that's not representative.

Reply:
I will when I get a chance but not soon I'm afraid.

Reply:
I did some more testing and have now merged this. Thanks for your input and help!